nautobot/extras/models/groups.py

```diff
@@ -431,3 +441,155 @@ def clean(self):
+
+        return q
+
+    def process_group_filters(self, qs=None, group=None):
+        # Start with this group's base queryset
+        base_qs = self.get_queryset().clone()
+        if qs is None:
+            # Clone the base queryset so we can modify it.
+            qs = self.get_queryset().clone()
+
+        if group is None:
+            group = self
+            print(f"Processing group {group}...")
+
+        big_q = models.Q()
+
+        # Enumerate the filters. Recursing into any groups of groups.
+        for membership in group.dynamic_group_memberships.all():
+            group = membership.group
+            print(f"Processing group {group}...")
+            child_memberships = group.dynamic_group_memberships.all()
+            if child_memberships.exists():
+                qs = self.process_group_filters(qs, group)
+
+            operator = membership.operator
+
+            print(f"{group} -> {group.filter} -> {operator}")
+
+            # The previous iteration that used set operations at the DB layer
+            # but limited by not being able to do further filtering.
+            """
+            fs = self.filterset_class(group.filter, self.get_queryset())
+            if operator == "union":
+                qs = qs.union(fs.qs)
+            elif operator == "difference":
+                qs = qs.difference(fs.qs)
+            elif operator == "intersection":
+                qs = qs.intersection(fs.qs)
+            """
+            next_set = self.generate_query_for_group(group)
+            # print(next_set)
+            if operator == "union":
+                big_q |= next_set
+                qs = qs | base_qs.filter(next_set)
+            elif operator == "difference":
+                big_q &= ~next_set
+                qs = qs.exclude(next_set)
+            elif operator == "intersection":
+                big_q &= next_set
+                qs = qs.filter(next_set)
+            # print(big_q)
+
+        # return big_q
+        return qs
+
+
+class DynamicGroupFilterActionChoices(models.TextChoices):
+    UNION = "union", "Union (OR)"
+    INTERSECTION = "intersection", "Intersection (AND)"
+    DIFFERENCE = "difference", "Difference (NOT)"
+
+
+class DynamicGroupMembership(BaseModel):
+    """Intermediate model for associating filters to groups."""
+
+    group = models.ForeignKey(
+        "extras.DynamicGroup",
+        on_delete=models.CASCADE,
+        related_name="+",
+    )
+    parent_group = models.ForeignKey(
+        "extras.DynamicGroup", on_delete=models.CASCADE, related_name="dynamic_group_memberships"
+    )
+    operator = models.CharField(choices=DynamicGroupFilterActionChoices.choices, max_length=12)
+    weight = models.PositiveSmallIntegerField()
+
+    def __str__(self):
+        return f"{self.group}: {self.operator} ({self.weight})"
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=["group", "parent_group"], name="group_to_filter_uniq"),
```

Comment:
There was a comment to drop this uniqueness constraint between `parent_group` and `group`, Such as if we ...

> might want or need to directly include the same child group more than once? (A intersection B union C difference B, or something)?

Reply:
We should probably add `weight` to this uniqueness constraint so that the same `group` can't be used with the same `parent_group` with the same `weight`.